spec/std/isa/csr/mtvec.yaml

```
@@ -21,12 +23,45 @@ fields:
       Bits [MXLEN-1:2] of the exception vector physical address for any trap taken in M-mode.
 
       The implementation physical memory map may resitrict which values are legal in this field.
-    type: RW-R
+    type(): |
+      return MTVEC_ACCESS == "read-only" ? CsrFieldType::RO : CsrFieldType::RWR;
     sw_write(csr_value): |
       # Base spec says that BASE must be 4-byte aligned, which will always be the case
       # implementations may put further constraints on BASE when MODE != Direct
       # If that is the case, stvec should have an override for the implementation
-      return csr_value.BASE;
+      if (csr_value.MODE == 0) {
+        if ($array_includes?(MTVEC_MODES, 0)) {
+          return csr_value.BASE;
+        } else {
+          if (MTVEC_ILLEGAL_WRITE_BEHAVIOR == "retain") {
+            return CSR[mtvec].BASE;
+          } else if (MTVEC_ILLEGAL_WRITE_BEHAVIOR == "custom") {
+            return UNDEFINED_LEGAL_DETERMINISTIC;
+          } else {
+            unreachable();
+          }
+        }
+      } else if (csr_value.MODE == 1) {
+        if ($array_includes?(MTVEC_MODES, 1)) {
+          return csr_value.BASE;
+        } else {
+          if (MTVEC_ILLEGAL_WRITE_BEHAVIOR == "retain") {
+            return CSR[mtvec].BASE;
+          } else if (MTVEC_ILLEGAL_WRITE_BEHAVIOR == "custom") {
+            return UNDEFINED_LEGAL_DETERMINISTIC;
+          } else {
+            unreachable();
+          }
+        }
+      } else {
+        if (MTVEC_ILLEGAL_WRITE_BEHAVIOR == "retain") {
+          return CSR[mtvec].BASE;
+        } else if (MTVEC_ILLEGAL_WRITE_BEHAVIOR == "custom") {
+          return UNDEFINED_LEGAL_DETERMINISTIC;
+        } else {
+          unreachable();
+        }
+      }
```

Nit, but you have a bit more structure here than needed:
```suggestion
      if (csr_value.MODE == 0) {
        if ($array_includes?(MTVEC_MODES, 0)) {
          return csr_value.BASE;
        }
        if (MTVEC_ILLEGAL_WRITE_BEHAVIOR == "retain") {
          return CSR[mtvec].BASE;
        }
        if (MTVEC_ILLEGAL_WRITE_BEHAVIOR == "custom") {
          return UNDEFINED_LEGAL_DETERMINISTIC;
        }
        unreachable();
      } else if (csr_value.MODE == 1) {
        if ($array_includes?(MTVEC_MODES, 1)) {
          return csr_value.BASE;
        }
        if (MTVEC_ILLEGAL_WRITE_BEHAVIOR == "retain") {
          return CSR[mtvec].BASE;
        }
        if (MTVEC_ILLEGAL_WRITE_BEHAVIOR == "custom") {
          return UNDEFINED_LEGAL_DETERMINISTIC;
        }
        unreachable();
      } else {
        if (MTVEC_ILLEGAL_WRITE_BEHAVIOR == "retain") {
          return CSR[mtvec].BASE;
        }
        if (MTVEC_ILLEGAL_WRITE_BEHAVIOR == "custom") {
          return UNDEFINED_LEGAL_DETERMINISTIC;
        }
        unreachable();
      }
```